os/kernel/fault_manager/fault_manager.c

```diff
@@ -0,0 +1,297 @@
+void fm_printbinarytable(void)
+{
+	int i;
+
+	lldbg("-------------------------------------\n");
+	lldbg("BIN_ID	BIN_PATH\n");
+	for (i = 0; i < CONFIG_NUM_BINARIES; i++) {
+		lldbg("%d\t%s\n", reg_table[i].binid, reg_table[i].bin_path);
+	}
+	lldbg("-------------------------------------\n");
+
+}
+
+/****************************************************************************
+ * Name: get_binid
+ *
+ * Description:
+ *   This function will traverse the task/thread tree and return the root bin pid.
+ *   If fault happened in any of the thread/task, it will call this function to get
+ *   root binary pid for the same.
+ *
+ * Input parameters:
+ *   pid   -   The pid of the task/thread, whose root bin pid is required
+ *
+ * Returned Value:
+ *   pid   -   pid of the main binary of input pid
+ *
+ ****************************************************************************/
+pid_t get_binid(pid_t pid)
+{
+	/* TODO: Implementation of traverser of the tree to get main binary pid */
+	return pid;
+}
+
+/****************************************************************************
+ * Name: kill_binary
+ *
+ * Description:
+ *   This function will kill all the task/thread created by the binary
+ *   i.e input pid.
+ *   If any fault occurs in the system, fault manager will call this function to
+ *   kill all the thread/task created by the input binary.
+ *
+ * Input parameters:
+ *   pid   -   The pid of the binary, whoes all childs to be killed
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ ****************************************************************************/
+int kill_binary(pid_t pid)
+{
+	int ret;
+
+	/* TODO: Implementation of traverser of tree and kill all the task/thread
+	 * of the main binary
+	 */
+	ret = kill(pid, SIGWORK);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: fault_manager
+ *
+ * Description:
+ *   This function will receive the faulty binary's pid and check if it is one
+ *   of the registered binary with fault manager.
+ *   If the binary is registered then it will be restarted after killing its
+ *   child processes, and if it is not registered then board will be rebooted.
+ *
+ * Input parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ * Assumptions:
+ *    All the binaries of the system will be registered with the fault manager
+ *
+ ****************************************************************************/
+int fault_manager(int argc, char *argv[])
+{
+	mqd_t mqdes;
+	int ret;
+	int i;
+	int index = -1;
+	pid_t pid;
+	pid_t pid_restart;
+	pid_t bin_id;
+	struct mq_attr attr;
+	char restart_bin[PATH_LEN] = " ";
+
+	/* Initialize the binary table with default value */
+	for (i = 0; i < CONFIG_NUM_BINARIES; i++) {
+		reg_table[i].binid = -1;
+		reg_table[i].bin_path[0] = '\0';
+	}
+
+	attr.mq_maxmsg = 32;
```

Comment:
Dont hardcode, use #define for maximum size

Reply:
Modified